File: Design War Measurement System/Code/Erlang/measure.erl
-module(measure).
-compile(export_all).

-define (MAX_INT, 16#7FFFFFFF).

%% Vector Operations
vec_equal(Vec1, Vec2) ->
    Vec1 =:= Vec2.

vec_add(Vec1, Vec2) ->
    case length(Vec1) =:= length(Vec2) of
        true ->
            lists:zipwith(fun(X, Y) -> X + Y end, Vec1, Vec2);
        false ->
            unequal_lengths
    end.

vec_dotproduct(Vec1, Vec2) ->
    case length(Vec1) =:= length(Vec2) of
        true ->
            lists:sum(lists:zipwith(fun(X, Y) -> X * Y end, Vec1, Vec2));
        false ->
            unequal_lengths
    end.

vec_div(Vec1, Vec2) ->
    case length(Vec1) =:= length(Vec2) andalso length(Vec2) =/= 0 andalso (not lists:member(0, Vec2)) of
        true ->
            vec_div_after_validate(lists:reverse(Vec1), lists:reverse(Vec2), [], 0);
        false ->
            case length(Vec2) =:= 0 of
                true ->
                    zero_divisor_vec;
                false ->
                    case lists:member(0, Vec2) of
                        true ->
                            zero_divisor;
                        false ->
                            unequal_lengths
                    end
            end
    end.

vec_div_after_validate([H_RevVec1], _Vec2, ResVec, Carry) ->
    [Carry + H_RevVec1|ResVec];
vec_div_after_validate([H_RevVec1|T_RevVec1], [H_Vec2|T_Vec2], ResVec, Carry) ->
    Dividend = H_RevVec1 + Carry,
    Remainder = Dividend rem H_Vec2,
    vec_div_after_validate(T_RevVec1, T_Vec2, [Remainder|ResVec], Dividend div H_Vec2).

vec_dotdiv(BaseValue, Vec) ->
    case length(Vec) =/= 0 andalso (not lists:member(0, Vec)) of
        true ->
            vec_dotdiv_after_validate(BaseValue, Vec, []);
        false ->
            case length(Vec) =:= 0 of
                true ->
                    zero_divisor_vec;
                false ->
                    case lists:member(0, Vec) of
                        true ->
                            zero_divisor;
                        false ->
                            unequal_lengths
                    end
            end
    end.

vec_dotdiv_after_validate(0, [], ResVec) ->
    lists:reverse(ResVec);
vec_dotdiv_after_validate(Value, [H_Vec|T_Vec], ResVec) ->
    vec_dotdiv_after_validate(Value rem H_Vec, T_Vec, [Value div H_Vec|ResVec]).

%% Measurement System
ms_base_factors(StepFactors) ->
    ms_base_factors(StepFactors, []).

ms_base_factors([], BaseFactors) ->
    lists:reverse(BaseFactors);
ms_base_factors([_H_HSF|T_HSF], BaseFactors) ->
    ms_base_factors(T_HSF, [lists:foldl(fun(X, Y) -> X * Y end, 1, T_HSF)|BaseFactors]).

ms_base(QuantityVec, BaseFactors) ->
    vec_dotproduct(QuantityVec, BaseFactors).

ms_equal(QuantityVec1, QuantityVec2, StepFactors) ->
    vec_equal(ms_normalize(QuantityVec1, StepFactors), ms_normalize(QuantityVec2, StepFactors)).

ms_normalize(QuantityVec, StepFactors) ->
    vec_div(QuantityVec, StepFactors).

ms_add(QuantityVec1, QuantityVec2, StepFactors) ->
    ms_normalize(vec_add(QuantityVec1, QuantityVec2), StepFactors).

%% Measurement System UI
ms_ui_parse_unitconversiondesc(Desc) ->
    Tokens = string:tokens(string:strip(Desc), " +"),
    ms_ui_parse_unitconversiondesc(Tokens, [?MAX_INT], []).

ms_ui_parse_unitconversiondesc([BaseUnit], StepFactors, SysUnits) ->
    {lists:reverse(StepFactors), lists:reverse([BaseUnit|SysUnits])};
ms_ui_parse_unitconversiondesc([H1_Tokens, H2_Tokens|T_Tokens], StepFactors, SysUnits) ->
    ms_ui_parse_unitconversiondesc(T_Tokens, [list_to_integer(H2_Tokens)|StepFactors], [H1_Tokens|SysUnits]).

ms_ui_parse_quantity(Quantity, SysUnits) ->
    Tokens = string:tokens(string:strip(Quantity), " +"),
    {QuantityValues, QuantityUnits} = ms_ui_parse_quantity_values_and_units(Tokens, [], []),
    ms_ui_quantity_vec(QuantityValues, QuantityUnits, SysUnits, 1, []).

ms_ui_parse_quantity_values_and_units([], QuantityValues, QuantityUnits) ->
    {lists:reverse(QuantityValues), lists:reverse(QuantityUnits)};
ms_ui_parse_quantity_values_and_units([H1_Tokens, H2_Tokens|T_Tokens], QuantityValues, QuantityUnits) ->
    ms_ui_parse_quantity_values_and_units(T_Tokens, [list_to_integer(H1_Tokens)|QuantityValues], [H2_Tokens|QuantityUnits]).

ms_ui_quantity_vec(_QuantityValues, _QuantityUnits, [], _Idx, QuantityVec) ->
    lists:reverse(QuantityVec);
ms_ui_quantity_vec(QuantityValues, QuantityUnits, [H_SysUnits|T_SysUnits], Idx, QuantityVec) ->
    case lists:member(H_SysUnits, QuantityUnits) of
        true ->
            ms_ui_quantity_vec(QuantityValues, QuantityUnits, T_SysUnits, Idx + 1, [lists:nth(Idx, QuantityValues)|QuantityVec]);
        false ->
            ms_ui_quantity_vec(QuantityValues, QuantityUnits, T_SysUnits, Idx, [0|QuantityVec])
    end.

ms_ui_equal(Quantity1, Quantity2, SysUnits, StepFactors) ->
    ms_equal(ms_ui_parse_quantity(Quantity1, SysUnits), ms_ui_parse_quantity(Quantity2, SysUnits), StepFactors).

ms_ui_format(QuantityVec, SysUnits) ->
    ms_ui_format(QuantityVec, SysUnits, "").

ms_ui_format([], [], [_|Res]) ->
    Res;
ms_ui_format([H_QuantityVec|T_QuantityVec]=QuantityVec, [H_SysUnits|T_SysUnits], Res) ->
    case H_QuantityVec =/= 0 orelse length(QuantityVec) =:= 1 of
        true ->
            ValueAndUnit = string:join([integer_to_list(H_QuantityVec), H_SysUnits], " "),
            NewRes = string:join([Res, ValueAndUnit], " "),
            ms_ui_format(T_QuantityVec, T_SysUnits, NewRes);
        false ->
            ms_ui_format(T_QuantityVec, T_SysUnits, Res)
    end.

ms_ui_add(Quantity1, Quantity2, SysUnits, StepFactors) ->
    ms_ui_format(ms_add(ms_ui_parse_quantity(Quantity1, SysUnits), ms_ui_parse_quantity(Quantity2, SysUnits), StepFactors), SysUnits).

ms_ui_baseformat(QuantityVec, SysUnits, StepFactors) ->
    BaseFactors = ms_base_factors(StepFactors),
    BasedValue = ms_base(QuantityVec, BaseFactors),
    string:join([integer_to_list(BasedValue), lists:last(SysUnits)], " ").

%% Tests
% Vector Operation Tests
test_vec_equal() ->
    true = vec_equal([1, 2, 3], [1, 2, 3]),
    true = vec_equal([], []),
    false = vec_equal([1, 2], [1, 2, 3]),
    false = vec_equal([1, 3, 4], [1, 2, 3]),

    test_vec_equal_ok.

test_vec_add() ->
    [5, 7, 9] = vec_add([1, 2, 3], [4, 5, 6]),
    [] = vec_add([], []),
    unequal_lengths = vec_add([1, 2], [4, 5, 6]),

    test_vec_add_ok.

test_vec_dotproduct() ->
    102 = vec_dotproduct([1, 0, 2], [100, 10, 1]),
    0 = vec_dotproduct([], []),
    unequal_lengths = vec_dotproduct([1, 2], [4, 5, 6]),

    test_vec_dotproduct_ok.

test_vec_div() ->
    [0, 0, 2, 0] = vec_div([0, 0, 0, 24], [?MAX_INT, 1760, 3, 12]),
    [2, 6, 0, 4] = vec_div([1, 1765, 0, 40], [?MAX_INT, 1760, 3, 12]),
    unequal_lengths = vec_div([1, 2, 3], [1, 2]),
    zero_divisor_vec = vec_div([], []),
    zero_divisor = vec_div([1, 2], [0, 4]),

    test_vec_div_ok.

test_vec_dotdiv() ->
    [1, 0, 2] = vec_dotdiv(102, [100, 10, 1]),
    zero_divisor_vec = vec_dotdiv(0, []),
    zero_divisor = vec_dotdiv(0, [0, 4]),

    test_vec_dotdiv_ok.

% Measurement System Tests
test_ms_base_factors() ->
    [1760 * 3 * 12, 3 * 12, 12, 1] = ms_base_factors([?MAX_INT, 1760, 3, 12]),

    test_ms_base_factors_ok.

test_ms_base() ->
    BaseFactors = [1760 * 3 * 12, 3 * 12, 12, 1],
    63472 = ms_base([1, 2, 3, 4], BaseFactors),
    63396 = ms_base([1, 0, 3, 0], BaseFactors),

    test_ms_base_ok.

test_ms_equal() ->
    StepFactors = [?MAX_INT, 1760, 3, 12],
    true = ms_equal([1, 2, 3, 4], [0, 0, 0, 63472], StepFactors),
    true = ms_equal([0, 1765, 0, 40], [1, 6, 0, 4], StepFactors),
    false = ms_equal([0, 1765, 0, 41], [1, 6, 0, 4], StepFactors),

    test_ms_equal_ok.

test_ms_normalize() ->
    StepFactors = [?MAX_INT, 1760, 3, 12],
    [1, 4, 0, 1] = ms_normalize([0, 1762, 5, 13], StepFactors),

    test_ms_normalize_ok.

test_ms_add() ->
    StepFactors = [?MAX_INT, 1760, 3, 12],
    [0, 0, 2, 0] = ms_add([0, 0, 0, 13], [0, 0, 0, 11], StepFactors),
    [0, 3, 0, 0] = ms_add([0, 0, 3, 0], [0, 2, 0, 0], StepFactors),

    test_ms_add_ok.

% Measurement System UI Tests
test_ms_ui_parse_unitconversiondesc() ->
    StepFactors = [?MAX_INT, 1760, 3, 12],
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    {StepFactors, SysUnits} = ms_ui_parse_unitconversiondesc(" Mile  1760 Yard 3 Feet 12 Inch  ").

test_ms_ui_parse_quantity() ->
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    [1, 2, 3, 4] = ms_ui_parse_quantity(" 1 Mile  2 Yard 3 Feet 4 Inch  ", SysUnits),
    % [1, 2, 0, 4] = ms_ui_parse_quantity("1 Mile 4 Inch 2 Yard", SysUnits), %not supported temporarily
    [1, 2, 0, 4] = ms_ui_parse_quantity("1 Mile 2 Yard 4 Inch", SysUnits),
    [0, 0, 3, 0] = ms_ui_parse_quantity("3 Feet", SysUnits),
    
    test_ms_ui_parse_quantity_ok.

test_ms_ui_equal() ->
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    StepFactors = [?MAX_INT, 1760, 3, 12],
    true = ms_ui_equal("1 Mile 2 Yard 3 Feet 4 Inch", "63472 Inch", SysUnits, StepFactors),
    true = ms_ui_equal("1765 Yard 40 Inch", "1 Mile 6 Yard 4 Inch", SysUnits, StepFactors),
    false = ms_ui_equal("1765 Yard 41 Inch", "1 Mile 6 Yard 4 Inch", SysUnits, StepFactors),

    test_ms_ui_equal_ok.

test_ms_ui_format() ->
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    "1 Mile 3 Yard 4 Inch" = ms_ui_format([1, 3, 0, 4], SysUnits),
    "2 Feet 0 Inch" = ms_ui_format([0, 0, 2, 0], SysUnits),

    test_ms_ui_format_ok.

test_ms_ui_add() ->
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    StepFactors = [?MAX_INT, 1760, 3, 12],
    "2 Feet 0 Inch" = ms_ui_add("13 Inch", "11 Inch", SysUnits, StepFactors),
    "3 Yard 0 Inch" = ms_ui_add("3 Feet", "2 Yard", SysUnits, StepFactors),

    test_ms_ui_add_ok.

test_baseFormat() ->
    SysUnits = ["Mile", "Yard", "Feet", "Inch"],
    StepFactors = [?MAX_INT, 1760, 3, 12],
    "63472 Inch" = ms_ui_baseformat([1, 3, 0, 4], SysUnits, StepFactors),
    "2 Feet 0 Inch" = ms_ui_add("13 Inch", "11 Inch", SysUnits, StepFactors),
    
    test_baseFormat_ok.     

test() ->
    test_vec_equal(),
    test_vec_add(),
    test_vec_dotproduct(),
    test_vec_div(),
    test_vec_dotdiv(),

    test_ms_base_factors(),
    test_ms_base(),
    test_ms_equal(),
    test_ms_normalize(),
    test_ms_add(),

    test_ms_ui_parse_unitconversiondesc(),
    test_ms_ui_parse_quantity(),
    test_ms_ui_format(),
    test_ms_ui_add(),
    test_baseFormat(),

    test_ok.

main() ->
    {StepFactors, SysUnits} = ms_ui_parse_unitconversiondesc("Mile 1760 Yard 3 Feet 12 Inch"),
    R1 = ms_ui_equal("1765 Yard 40 Inch", "1 Mile 6 Yard 4 Inch", SysUnits, StepFactors),
    io:format("1765 Yard 40 Inch == 1 Mile 6 Yard 4 Inch ? ~p~n", [R1]),
    R2 = ms_ui_add("13 Inch", "11 Inch", SysUnits, StepFactors),
    io:format("13 Inch + 11 Inch = ~p~n", [R2]),

    {StepFactors2, SysUnits2} = ms_ui_parse_unitconversiondesc("OZ 2 TBSP 3 TSP"),
    R3 = ms_ui_equal("1 OZ 10 TSP", "2 OZ 1 TBSP 1 TSP", SysUnits2, StepFactors2),
    io:format("1 OZ 10 TSP == 2 OZ 1 TBSP 1 TSP ? ~p~n", [R3]),
    R4 = ms_ui_add("1 OZ", "3 TBSP 3 TSP", SysUnits2, StepFactors2),
    io:format("1 OZ + 3 TBSP 3 TSP = ~p~n", [R4]),

    {StepFactors3, SysUnits3} = ms_ui_parse_unitconversiondesc("Day 24 Hour 60 Minute 60 Second"),
    R5 = ms_ui_equal("7 Hour 59 Minute 60 Second", "8 Hour", SysUnits3, StepFactors3),
    io:format("7 Hour 59 Minute 60 Second == 8 Hour ? ~p~n", [R5]),
    R6 = ms_ui_add("1 Day 23 Hour 59 Minute 59 Second", "1 Second", SysUnits3, StepFactors3),
    io:format("1 Day 23 Hour 59 Minute 59 Second + 1 Second = ~p~n", [R6]).

File: DDD DSL Design War Measurement System/Code/Python/measure.py
## Vector Operations
def vec_equal(vec1, vec2):
    return vec1 == vec2

def vec_add(vec1, vec2):
    if len(vec1) == len(vec2):
        return [c1 + c2 for (c1, c2) in zip(vec1, vec2)]
    else:
        return None

def vec_dotproduct(vec1, vec2):
    if len(vec1) == len(vec2):
        return sum([c1 * c2 for (c1, c2) in zip(vec1, vec2)])
    else:
        return None

def vec_div(vec1, vec2):
    if len(vec1) == len(vec2) and len(vec2) != 0 and (0 not in vec2):
        return vec_div_after_validation(vec1[::-1], vec2[::-1])
    else:
        return None

def vec_div_after_validation(rvec1, rvec2):
    result = []
    carry = 0
    for (c1, c2) in zip(rvec1, rvec2):
        dividend = c1 + carry
        divisor = c2
        remainder = dividend % divisor
        result.append(remainder)
        carry = dividend / divisor
    return result[::-1]

def vec_dotdiv(baseValue, vec):
    if len(vec) != 0 and (0 not in vec):
        return vec_dotdiv_after_validation(baseValue, vec)
    else:
        return None

def vec_dotdiv_after_validation(baseValue, vec):
    result = []
    remainder = baseValue
    for c in vec:
        quotient = remainder / c
        remainder = remainder % c
        result.append(quotient)
    return result

def vec_scale(c, vec):
    return [c * vi for vi in vec]

## Measurement System
def ms_base_factors(stepFactors):
    result = []
    for (i, stepFactor) in enumerate(stepFactors[1:]):
        result.append(reduce(lambda x, y: x * y, stepFactors[i + 1:]))
    result.append(1)
    return result

def ms_base(quantityVec, baseFactors):
    return vec_dotproduct(quantityVec, baseFactors)

def ms_normalize(quantityVec, stepFactors):
    return vec_div(quantityVec, stepFactors)

def ms_equal(quantityVec1, quantityVec2, stepFactors):
    return vec_equal(ms_normalize(quantityVec1, stepFactors), ms_normalize(quantityVec2, stepFactors))

def ms_add(quantityVec1, quantityVec2, stepFactors):
    return ms_normalize(vec_add(quantityVec1, quantityVec2), stepFactors)

def ms_scale(c, quantityVec, stepFactors):
    return ms_normalize(vec_scale(c, quantityVec), stepFactors)

## Measurement System UI
import re
import string

def ms_ui_parse_unitconversiondesc(desc):
    tokens = re.split(' +', string.strip(desc))
    stepFactors = [MAX_INT()]
    sysUnits = [tokens[0]]
    for i in range(1, len(tokens), 2):
        stepFactors.append(int(tokens[i]))
        sysUnits.append(tokens[i + 1])
    return (stepFactors, sysUnits)

def ms_ui_parse_quantity(quantity, sysUnits):
    (values, units) = ms_ui_parse_values_and_units(quantity)
    return [ms_ui_quantity_value(values, units, sysUnit) for sysUnit in sysUnits]

def ms_ui_parse_values_and_units(quantity):
    tokens = re.split(' +', string.strip(quantity))
    values = []
    units = []
    for i in range(0, len(tokens), 2):
        values.append(int(tokens[i]))
        units.append(tokens[i + 1])
    return (values, units)

def ms_ui_quantity_value(values, units, sysUnit):
    if sysUnit in units:
        return values[units.index(sysUnit)]
    return 0        

def ms_ui_equal(quantity1, quantity2, sysUnits, stepFactors):
    return ms_equal(ms_ui_parse_quantity(quantity1, sysUnits), 
                    ms_ui_parse_quantity(quantity2, sysUnits), stepFactors)

def ms_ui_format(quantityVec, sysUnits):
    result = ''
    for i in xrange(0, len(quantityVec)):
        if quantityVec[i] != 0 or i == len(quantityVec) - 1:
            valueAndUnit = ' '.join([str(quantityVec[i]), sysUnits[i]])
            result = ' '.join([result, valueAndUnit])
    return result[1:]

def ms_ui_add(quantity1, quantity2, sysUnits, stepFactors):
    return ms_ui_format(ms_add(ms_ui_parse_quantity(quantity1, sysUnits), 
                               ms_ui_parse_quantity(quantity2, sysUnits), stepFactors), sysUnits)

def ms_ui_baseformat(quantityVec, sysUnits, stepFactors):
    baseFactors = ms_base_factors(stepFactors)
    basedValue = ms_base(quantityVec, baseFactors)
    return (' '.join([str(basedValue), sysUnits[-1]]))

def ms_ui_scale(c, quantity, sysUnits, stepFactors):
    return ms_ui_format(ms_scale(int(c), ms_ui_parse_quantity(quantity, sysUnits), stepFactors), sysUnits)

def MAX_INT():
    return 0x7FFFFFFF

## Tests
# Vector Operation Tests
def test_vec_equal():
    assert vec_equal([1, 2, 3], [1, 2, 3])
    assert vec_equal([], [])
    assert not vec_equal([1, 2], [1, 2, 3])
    assert not vec_equal([1, 3, 4], [1, 2, 3])

    print 'test_vec_equal ok'

def test_vec_add():
    assert [5, 7, 9] == vec_add([1, 2, 3], [4, 5, 6])
    assert [] == vec_add([], [])
    assert None == vec_add([1, 2], [4, 5, 6])

    print 'test_vec_add ok'

def test_vec_dotproduct():
    assert 102 == vec_dotproduct([1, 0, 2], [100, 10, 1])
    assert 0 == vec_dotproduct([], [])
    assert None == vec_dotproduct([1, 2], [4, 5, 6])

    print 'test_vec_dotproduct ok'

def test_vec_div():
    assert [0, 0, 2, 0] == vec_div([0, 0, 0, 24], [MAX_INT(), 1760, 3, 12])
    assert [2, 6, 0, 4] == vec_div([1, 1765, 0, 40], [MAX_INT(), 1760, 3, 12])
    assert None == vec_div([1, 2, 3], [1, 2])
    assert None == vec_div([], [])
    assert None == vec_div([1, 2], [0, 4])

    print 'test_vec_div ok'

def test_vec_dotdiv():
    assert [1, 0, 2] == vec_dotdiv(102, [100, 10, 1])
    assert None == vec_dotdiv(0, [])
    assert None == vec_dotdiv(0, [0, 4])

    print 'test_vec_dotdiv ok'

def test_vec_scale():
    assert [2, 4, 6] == vec_scale(2, [1, 2, 3])
    assert [0, 0] == vec_scale(0, [1, 2])
    assert [] == vec_scale(2, [])

    print 'test_vec_scale ok'

# Measurement System Tests
def test_ms_base_factors():
    assert [1760 * 3 * 12, 3 * 12, 12, 1] == ms_base_factors([MAX_INT(), 1760, 3, 12])

    print 'test_ms_base_factors ok'

def test_ms_base():
    baseFactors = [1760 * 3 * 12, 3 * 12, 12, 1]
    assert 63472 == ms_base([1, 2, 3, 4], baseFactors)
    assert 63396 == ms_base([1, 0, 3, 0], baseFactors)

    print 'test_ms_base ok'

def test_ms_normalize():
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert [1, 4, 0, 1] == ms_normalize([0, 1762, 5, 13], stepFactors)

    print 'test_ms_normalize ok'

def test_ms_equal():
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert ms_equal([1, 2, 3, 4], [0, 0, 0, 63472], stepFactors)
    assert ms_equal([0, 1765, 0, 40], [1, 6, 0, 4], stepFactors)
    assert not ms_equal([0, 1765, 0, 41], [1, 6, 0, 4], stepFactors)

    print 'test_ms_equal ok'

def test_ms_add():
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert [0, 0, 2, 0] == ms_add([0, 0, 0, 13], [0, 0, 0, 11], stepFactors)
    assert [0, 3, 0, 0] == ms_add([0, 0, 3, 0], [0, 2, 0, 0], stepFactors)

    print 'test_ms_add ok'

def test_ms_scale():
    stepFactors = [MAX_INT(), 1760, 3, 12]

    assert [0, 6, 0, 0] == ms_scale(2, [0, 2, 3, 0], stepFactors)
    assert [0, 6, 1, 1] == ms_add(ms_scale(2, [0, 2, 3, 0], stepFactors), [0, 0, 0, 13], stepFactors)

    print 'test_ms_scale ok'

# Measurement System UI Tests
def test_ms_ui_parse_unitconversiondesc():
    stepFactors = [MAX_INT(), 1760, 3, 12]
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    assert (stepFactors, sysUnits) == ms_ui_parse_unitconversiondesc(' Mile  1760 Yard 3 Feet 12 Inch  ')

    print 'test_ms_ui_parse_unitconversiondesc ok'

def test_ms_ui_parse_quantity():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    assert [1, 2, 3, 4] == ms_ui_parse_quantity(' 1 Mile  2 Yard 3 Feet 4 Inch  ', sysUnits)
    assert [1, 2, 0, 4] == ms_ui_parse_quantity('1 Mile 4 Inch 2 Yard', sysUnits)
    assert [1, 2, 0, 4] == ms_ui_parse_quantity('1 Mile 2 Yard 4 Inch', sysUnits)
    assert [0, 0, 3, 0] == ms_ui_parse_quantity('3 Feet', sysUnits)
    
    print 'test_ms_ui_parse_quantity ok'

def test_ms_ui_equal():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert ms_ui_equal('1 Mile 2 Yard 3 Feet 4 Inch', '63472 Inch', sysUnits, stepFactors)
    assert ms_ui_equal('1765 Yard 40 Inch', '1 Mile 6 Yard 4 Inch', sysUnits, stepFactors)
    assert not ms_ui_equal('1765 Yard 41 Inch', '1 Mile 6 Yard 4 Inch', sysUnits, stepFactors)

    print 'test_ms_ui_equal ok'

def test_ms_ui_format():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    assert '1 Mile 3 Yard 4 Inch' == ms_ui_format([1, 3, 0, 4], sysUnits)
    assert '2 Feet 0 Inch' == ms_ui_format([0, 0, 2, 0], sysUnits)

    print 'test_ms_ui_format ok'

def test_ms_ui_add():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert '2 Feet 0 Inch' == ms_ui_add('13 Inch', '11 Inch', sysUnits, stepFactors)
    assert '3 Yard 0 Inch' == ms_ui_add('3 Feet', '2 Yard', sysUnits, stepFactors)

    print 'test_ms_ui_add ok'

def test_ms_ui_baseFormat():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert '63472 Inch' == ms_ui_baseformat([1, 3, 0, 4], sysUnits, stepFactors)
    assert '2 Feet 0 Inch' == ms_ui_add('13 Inch', '11 Inch', sysUnits, stepFactors)
    
    print 'test_ms_ui_baseFormat ok'

def test_ms_ui_scale():
    sysUnits = ['Mile', 'Yard', 'Feet', 'Inch']
    stepFactors = [MAX_INT(), 1760, 3, 12]
    assert "6 Yard 0 Inch" == ms_ui_scale('2', '2 Yard 3 Feet', sysUnits, stepFactors)
    assert '6 Yard 1 Feet 1 Inch' == ms_ui_add(ms_ui_scale('2', '2 Yard 3 Feet', sysUnits, stepFactors), 
                                               '13 Inch', sysUnits, stepFactors)

    print 'test_ms_ui_scale ok'

def test():
    test_vec_equal()
    test_vec_add()
    test_vec_dotproduct()
    test_vec_div()
    test_vec_dotdiv()
    test_vec_scale()

    test_ms_base_factors()
    test_ms_base()
    test_ms_normalize()
    test_ms_equal()
    test_ms_add()
    test_ms_scale()

    test_ms_ui_parse_unitconversiondesc()
    test_ms_ui_parse_quantity()
    test_ms_ui_equal()
    test_ms_ui_format()
    test_ms_ui_add()
    test_ms_ui_baseFormat()
    test_ms_ui_scale()

    print 'test ok'

def main():
    (stepFactors, sysUnits) = ms_ui_parse_unitconversiondesc('Mile 1760 Yard 3 Feet 12 Inch')
    r1 = ms_ui_equal("1765 Yard 40 Inch", "1 Mile 6 Yard 4 Inch", sysUnits, stepFactors),
    print '1765 Yard 40 Inch == 1 Mile 6 Yard 4 Inch ? %s' % (r1)
    r2 = ms_ui_add('13 Inch', '11 Inch', sysUnits, stepFactors)
    print '13 Inch + 11 Inch = %s' % (r2)
    r3 = ms_ui_scale('2', '2 Yard 3 Feet', sysUnits, stepFactors)
    print '2 * 2 Yard 3 Feet = %s' % (r3)
    r4 = ms_ui_add(ms_ui_scale('2', '2 Yard 3 Feet', sysUnits, stepFactors), '13 Inch', sysUnits, stepFactors)
    print '2 * 2 Yard 3 Feet + 13 Inch = %s' % (r4)

    (stepFactors2, sysUnits2) = ms_ui_parse_unitconversiondesc('OZ 2 TBSP 3 TSP')
    r5 = ms_ui_equal('1 OZ 10 TSP', '2 OZ 1 TBSP 1 TSP', sysUnits2, stepFactors2)
    print '1 OZ 10 TSP == 2 OZ 1 TBSP 1 TSP ? %s' % (r5)
    r6 = ms_ui_add('1 OZ', '3 TBSP 3 TSP', sysUnits2, stepFactors2)
    print '1 OZ + 3 TBSP 3 TSP = %s' % (r6)
    r7 = ms_ui_scale('3', '1 OZ 10 TSP', sysUnits2, stepFactors2)
    print '3 * 1 OZ 10 TSP = %s' % (r7)
    r8 = ms_ui_add(ms_ui_scale('3', '1 OZ 10 TSP', sysUnits2, stepFactors2), '3 TBSP 1 TSP', sysUnits2, stepFactors2)
    print '3 * 1 OZ 10 TSP + 3 TBSP 1 TSP = %s' % (r8)

    (stepFactors3, sysUnits3) = ms_ui_parse_unitconversiondesc('Day 24 Hour 60 Minute 60 Second')
    r9 = ms_ui_equal('7 Hour 59 Minute 60 Second', '8 Hour', sysUnits3, stepFactors3)
    print '7 Hour 59 Minute 60 Second == 8 Hour ? %s' % (r9)
    r10 = ms_ui_add('1 Day 23 Hour 59 Minute 59 Second', '1 Second', sysUnits3, stepFactors3)
    print '1 Day 23 Hour 59 Minute 59 Second + 1 Second = %s' % (r10)
    r11 = ms_ui_scale('2', '11 Hour 22 Minute 33 Second', sysUnits3, stepFactors3)
    print '2 * 11 Hour 22 Minute 33 Second = %s' % (r11)
    r12 = ms_ui_add(ms_ui_scale('2', '11 Hour 22 Minute 33 Second', sysUnits3, stepFactors3), '60 Second', sysUnits3, stepFactors3)
    print '2 * 11 Hour 22 Minute 33 Second + 60 Second = %s' % (r12)

    (stepFactors4, sysUnits4) = ms_ui_parse_unitconversiondesc('Mile 4 Miya 440 Yard 3 Feet 12 Inch')
    r13 = ms_ui_equal('4 Miya 446 Yard 40 Inch', '1 Mile 1 Miya 7 Yard 4 Inch', sysUnits4, stepFactors4)
    print '4 Miya 446 Yard 40 Inch == 1 Mile 1 Miya 7 Yard 4 Inch ? %s' % (r13)
    r14 = ms_ui_add('3 Miya 13 Inch', '445 Yard 11 Inch', sysUnits4, stepFactors4)
    print '3 Miya 13 Inch + 445 Yard 11 Inch = %s' % (r14)
    r15 = ms_ui_add(ms_ui_scale('2', '3 Miya 13 Inch', sysUnits4, stepFactors4), '445 Yard 11 Inch', sysUnits4, stepFactors4)
    print '2 * 3 Miya 13 Inch + 445 Yard 11 Inch = %s' % (r15)

if __name__ == "__main__":
    main()
